src/main/resources/db/changelog/alert/4.2.0/changelog-authentication.xml

```diff
@@ -0,0 +1,538 @@
+    <changeSet author="psantos" id="1556732237461-2">
+        <createTable tableName="ACCESS_OPERATIONS" schemaName="ALERT">
+            <column autoIncrement="true" name="ID" type="BIGINT(19)">
+                <constraints primaryKey="true" primaryKeyName="ACCESS_OPERATION_KEY"/>
+            </column>
+            <column name="OPERATION_NAME" type="VARCHAR(100)"/>
+        </createTable>
+    </changeSet>
+    <changeSet author="psantos" id="1556732237461-3">
+        <createTable tableName="PERMISSION_KEYS" schemaName="ALERT">
+            <column autoIncrement="true" name="ID" type="BIGINT(19)">
+                <constraints primaryKey="true" primaryKeyName="PERMISSION_KEYS_KEY"/>
+            </column>
+            <column name="KEY_NAME" type="VARCHAR(255)"/>
+        </createTable>
+    </changeSet>
+    <changeSet author="psantos" id="1556732237461-4">
+        <createTable tableName="PERMISSION_MATRIX" schemaName="ALERT">
+            <column name="ROLE_ID" type="BIGINT(19)">
+                <constraints nullable="false"/>
+            </column>
+            <column name="PERMISSION_KEY_ID" type="BIGINT(19)">
+                <constraints nullable="false"/>
+            </column>
+            <column name="ACCESS_OPERATION_ID" type="BIGINT(19)">
+                <constraints nullable="false"/>
+            </column>
+        </createTable>
+        <addPrimaryKey
+                schemaName="ALERT"
+                tableName="PERMISSION_MATRIX"
+                columnNames="ROLE_ID, PERMISSION_KEY_ID, ACCESS_OPERATION_ID"
+                constraintName="PERMISSION_MATRIX_KEY"
+        />
+        <addForeignKeyConstraint
+                baseTableSchemaName="ALERT"
+                baseTableName="PERMISSION_MATRIX"
+                baseColumnNames="ROLE_ID"
+                constraintName="FK_PERMISSION_ROLE"
+                referencedTableSchemaName="ALERT"
+                referencedTableName="ROLES"
+                referencedColumnNames="ID"
+                onDelete="CASCADE"
+        />
+        <addForeignKeyConstraint
+                baseTableSchemaName="ALERT"
+                baseTableName="PERMISSION_MATRIX"
+                baseColumnNames="PERMISSION_KEY_ID"
+                constraintName="FK_PERMISSION_KEY"
+                referencedTableSchemaName="ALERT"
+                referencedTableName="PERMISSION_KEYS"
+                referencedColumnNames="ID"
+                onDelete="CASCADE"
+        />
+        <addForeignKeyConstraint
+                baseTableSchemaName="ALERT"
+                baseTableName="PERMISSION_MATRIX"
+                baseColumnNames="ACCESS_OPERATION_ID"
+                constraintName="FK_PERMISSION_OPERATION"
+                referencedTableSchemaName="ALERT"
+                referencedTableName="ACCESS_OPERATIONS"
+                referencedColumnNames="ID"
+                onDelete="CASCADE"
+        />
+    </changeSet>
+    <changeSet author="psantos" id="1556732237461-5">
```

Comment:
I think some stored procedures could help readability here rather than 4 lines per insert with only one value changing. It might also help to split out table creation and table initialization.

Reply:
I will look into this.  I was thinking about it beforehand.  The matrix is slightly different between roles.